parity-crypto/src/extended.rs

```diff
@@ -0,0 +1,514 @@
+// Copyright 2015-2019 Parity Technologies (UK) Ltd.
+// This file is part of Parity Ethereum.
+
+// Parity Ethereum is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Ethereum is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Ethereum.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Extended keys
+
+use crate::{Secret, Public};
+use ethereum_types::H256;
+pub use self::derivation::Error as DerivationError;
+
+/// Represents label that can be stored as a part of key derivation
+pub trait Label {
+	/// Length of the data that label occupies
+	fn len() -> usize;
+
+	/// Store label data to the key derivation sequence
+	/// Must not use more than `len()` bytes from slice
+	fn store(&self, target: &mut [u8]);
+}
+
+impl Label for u32 {
+	fn len() -> usize { 4 }
+
+	fn store(&self, target: &mut [u8]) {
+		let bytes = self.to_be_bytes();
+		target[0..4].copy_from_slice(&bytes);
+	}
+}
+
+/// Key derivation over generic label `T`
+pub enum Derivation<T: Label> {
+	/// Soft key derivation (allow proof of parent)
+	Soft(T),
+	/// Hard key derivation (does not allow proof of parent)
+	Hard(T),
+}
+
+impl From<u32> for Derivation<u32> {
+	fn from(index: u32) -> Self {
+		if index < (2 << 30) {
```

Comment:
I'm also not a big fan of binary math, but here it makes sense, because of derivation definition. I've added more details about it here and for the whole module